internal/controllers/extension_controller.go

```diff
@@ -231,30 +241,61 @@ func (r *ExtensionReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-// TODO: follow up with mapping of all the available App statuses: https://github.com/carvel-dev/kapp-controller/blob/855063edee53315811a13ee8d5df1431ba258ede/pkg/apis/kappctrl/v1alpha1/status.go#L28-L35
-// mapAppStatusToInstalledCondition currently maps only the installed condition.
-func mapAppStatusToInstalledCondition(existingApp *kappctrlv1alpha1.App, ext *ocv1alpha1.Extension, bundleImage string) {
-	appReady := findStatusCondition(existingApp.Status.GenericStatus.Conditions, kappctrlv1alpha1.ReconcileSucceeded)
-	if appReady == nil {
-		ext.Status.InstalledBundleResource = ""
-		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "install status unknown", ext.Generation)
+// mapAppStatusToCondition maps the reconciling/deleting App conditions to the installed/deleting conditions on the Extension.
+func MapAppStatusToCondition(existingApp *kappctrlv1alpha1.App, ext *ocv1alpha1.Extension) {
+	// Note: App.Status.Inspect errors are never surfaced to App conditions, so are currently ignored when determining App status.
+	if ext == nil {
 		return
 	}
+	message := "app status unknown"
 
-	if appReady.Status != corev1.ConditionTrue {
-		ext.Status.InstalledBundleResource = ""
-		setInstalledStatusConditionFailed(
-			&ext.Status.Conditions,
-			appReady.Message,
-			ext.GetGeneration(),
-		)
+	if existingApp == nil {
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, message, ext.Generation)
 		return
 	}
+	message = existingApp.Status.FriendlyDescription
+	if strings.Contains(message, "Error (see .status.usefulErrorMessage for details)") || len(message) == 0 {
+		message = existingApp.Status.UsefulErrorMessage
+	}
+
+	orderedAppStatuses := []kappctrlv1alpha1.ConditionType{
+		kappctrlv1alpha1.DeleteFailed,
+		kappctrlv1alpha1.Deleting,
+		kappctrlv1alpha1.ReconcileSucceeded,
+		kappctrlv1alpha1.ReconcileFailed,
+		kappctrlv1alpha1.Reconciling,
+	}
+	appStatusMapFn := map[kappctrlv1alpha1.ConditionType]func(*[]metav1.Condition, string, int64){
+		kappctrlv1alpha1.DeleteFailed:       setProgressingStatusConditionDeleteFailed,
+		kappctrlv1alpha1.Deleting:           setProgressingStatusConditionDeleting,
+		kappctrlv1alpha1.ReconcileSucceeded: setProgressingStatusConditionReconcileSuccess,
+		kappctrlv1alpha1.ReconcileFailed:    setProgressingStatusConditionReconcileFailed,
+		kappctrlv1alpha1.Reconciling:        setProgressingStatusConditionReconciling,
+	}
+	for _, cond := range orderedAppStatuses {
+		if c := findStatusCondition(existingApp.Status.GenericStatus.Conditions, cond); c != nil && c.Status == corev1.ConditionTrue {
+			if len(message) == 0 {
+				message = c.Message
+			}
+			appStatusMapFn[cond](&ext.Status.Conditions, message, ext.Generation)
+			return
+		}
+	}
```

Comment:
Instead of looping through these `orderedAppStatuses` and then doing a loop through the conditions to find them, could we just loop through all the App conditions, check if there is a status mapping function, and if there is one call it?